File: src/jv/DotsAndSections.java
package jv;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;

/* Example
In:
4
4 7
1 3
2 5
5 6
Out:
2
3 6
*/
public class DotsAndSections {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int numOfSections = Integer.parseInt(br.readLine());
        ArrayList<List<Integer>> sections = new ArrayList<>();
        ArrayList<Integer> rights = new ArrayList<>();
        for (int i = 0; i < numOfSections; i++) {
            String[] input = br.readLine().split(" ");
            sections.add(Collections.singletonList(Integer.parseInt(input[0], Integer.parseInt(input[1]))));
        }

        sections.sort(Comparator.comparing(o -> o.get(1)));

        List<Integer> connected = new ArrayList<>();

        for (int i = 0; i < numOfSections; i++) {
            int right = sections.get(i).get(1);
            for (int j = 0; j < numOfSections; j++) {
                if (j != i && !connected.contains(i)) {
                    if (right >= sections.get(j).get(0) && right <= sections.get(j).get(1)) {
                        rights.add(right);
                        connected.add(j);
                    } else if (!connected.contains(j)) {
                        rights.add(right);
                        connected.add(i);
                    }
                }
            }
        }

        if (!connected.contains(numOfSections - 1))
            rights.add(sections.get(numOfSections - 1).get(1));
        System.out.println(rights.size());
        for (int el: rights) {
            System.out.print(el + " ");
        }
    }
}

File: src/kt/dotsAndSections.kt
package kt

/*
По данным n отрезкам необходимо найти множество точек минимального размера, для которого каждый из отрезков содержит хотя бы одну из точек.

В первой строке дано число 1≤n≤100 отрезков. Каждая из последующих n строк содержит по два числа 0≤l≤r≤109, задающих начало и конец отрезка. Выведите оптимальное число m точек и сами m точек. Если таких множеств точек несколько, выведите любое из них.

Sample Input 1:
3
1 3
2 5
3 6
Sample Output 1:
1
3
Sample Input 2:
4
4 7
1 3
2 5
5 6
Sample Output 2:
2
3 6
*/
fun main() {
    val numOfSections: Int = readLine()!!.toInt()
    val sections = ArrayList<List<Int>>()
    val rights = ArrayList<Int>()
    for (i in 0 until numOfSections) {
        sections.add(readLine()!!.split(" ").map { el -> el.toInt() })
    }
    sections.sortBy { it[1] }

    val connected = ArrayList<Int>()
    for (i in 0 until numOfSections) {
        val right = sections[i][1]
        for (j in 0 until numOfSections) {
            if (j != i && i !in connected) {
                if (right in sections[j][0]..sections[j][1]) {
                    rights.add(right)
                    connected.add(j)
                } else if (!connected.contains(j)) {
                    rights.add(right)
                    connected.add(i)
                }
            }
        }
    }

    if (!connected.contains(numOfSections - 1))
        rights.add(sections[numOfSections - 1][1])
    println("${rights.distinct().size}")
    rights.distinct().map { print("$it ") }
}